Grant Application - Reference Blockchain System

hoho-yin opened the pull request:
### Project Abstract

> Please provide a brief description of your project here summarising key points (1-2 paragraphs).
>
> If your application is a follow-up to a previous grant, please mention which one in the first line of the abstract and include a link to previous pull requests if applicable.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying?

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [ ] I have read and understood the [FAQs](https://github.com/w3f/Grants-Program/blob/master/docs/faq.md), [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/grant_guidelines_per_category.md) and [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/announcement-guidelines.md).
- [ ] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [ ] I have read and acknowledge the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [ ] The software delivered for this grant will be released under an open-source license specified in the application.
- [ ] The initial PR contains only one commit (squash and force-push if needed).
- [ ] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).
- [ ] I prefer the discussion of this application to be in a private Element/Matrix channel. My username is: ``@_______``

### How Did You Hear About our grants program?

- [ ] Social Media
- [ ] Hackathon
- [ ] Personal Recommendation
- [ ] Substrate Builders Program
- [ ] Investor/VC
- [ ] Online Search
- [ ] Other: _______

Noc2 reviewed (changes requested):
Thanks for the grant application. Something to be aware of is that we usually don’t support the development of solidity smart contracts. On the one hand it’s hard to know what already exists at this stage and on the other hand we are no solidity experts. We mostly focus on ink! Smart contracts and the development of pallets. Apart from this, I don’t understand why you want to develop a new token design as part of your first milestone? Why aren’t you using existing token standards for this? The “shares” you own for example is something you can easily track with any token based on the total available supply. In general, it would be nice if you could add more information to the milestones and add the headlines as deliverables. For example, “xyz design” is very unspecific and not something we can evaluate.

hoho-yin commented:
Dear @Noc2:
Hello!

Thanks for your kindly review.

> We mostly focus on ink! Smart contracts and the development of pallets. 
Yes we will develop our project in ink! Smart contracts and pallets. 

> Apart from this, I don’t understand why you want to develop a new token design as part of your first milestone? Why aren’t you using existing token standards for this?
2. We think designing a new token standard is needed because no existing token design has such reference function. In existing token, a new token couldn't refer to any old tokens, and share its earnings with all its referred tokens.

> The “shares” you own for example is something you can easily track with any token based on the total available supply. 
3. "Shares" mean the proportion of the revenue one could get from the earnings of the token representing the work, just like shares of one company in the stock market. For example if one work worths $10, and one person owns 10% shares of the token of the work, he earns $1. 

> In general, it would be nice if you could add more information to the milestones and add the headlines as deliverables. For example, “xyz design” is very unspecific and not something we can evaluate.
4. I have changed the headlines as specific deliverables. Details of the proposal could be discussed and improved after the grant application is approved.


Thanks a lot!

Noc2 reviewed (changes requested):
Thanks for the quick reply. However, the milestone table should actually contain the specific deliverables as number 1, 2, 3.. etc. including as many details as possible. We need to have these details before we sign the grant application. It’s not an option to add the details after we sign the grant. Also your grant still says “solidity smart contracts”. And in my opinion, you still don’t need to store the ownership/share percentage on-chain with each token. If all tokens are on-chain anyway as well as each interaction, you can calculate the ownership without specifically storing it for each token (and constantly updating it?). So, in my opinion there is a clear reason why no token exists that supports this functionality.

hoho-yin commented:
Dear @Noc2:
Hello!

Thanks for your kindly kindly review and golden feedback. I have revised according to your critiques. 

> However, the milestone table should actually contain the specific deliverables as number 1, 2, 3.. etc. including as many details as possible. 
Sorry for ignoring that mistake. I have changed the numbers and added more details to the milestone table.

>  Also your grant still says “solidity smart contracts”. 
I have changed it to !ink on Pallets. We have started learning !ink, and will develop on Pallets after getting the grant.

> And in my opinion, you still don’t need to store the ownership/share percentage on-chain with each token. If all tokens are on-chain anyway as well as each interaction, you can calculate the ownership without specifically storing it for each token (and constantly updating it?). So, in my opinion there is a clear reason why no token exists that supports this functionality.
Yes I agree and I have deleted the shares and investment part. Now our proposal more focus on the reference relationship among different works and their revenue distribution, rather than the distribution within each work. 

Thanks a lot!

Kindly Regards,
@hoho-yin

Noc2 reviewed (changes requested):
Thanks for pinging me. However, regarding the table I was referring to adding the extra deliverables as additional rows with numbers and not changing the default deliverables. See for example https://github.com/w3f/Grants-Program/blob/master/applications/perun_channels.md#milestone-1--setup--core-types--channel-opening

hoho-yin commented:
> Thanks for pinging me. However, regarding the table I was referring to adding the extra deliverables as additional rows with numbers and not changing the default deliverables. See for example https://github.com/w3f/Grants-Program/blob/master/applications/perun_channels.md#milestone-1--setup--core-types--channel-opening

Dear @Noc2:

Thanks for your kindly feedback. I have revised the milestone part based on the example you gave. If any other deliverables are need, feel free to specify and I add them. I have also revised other parts to make my proposal more clear and specific. Thanks a lot and look forward to your response!


Kindly Regards,
Jieliang Yin

Noc2 reviewed (changes requested):
Thanks for the update and the additional information. I will mark the application as ready for review, so that others start looking into it as well. But could you explain for example why you need to create the token standard, since you removed the part about “shares”?

laboon reviewed (commented):
I think at a very high level this is interesting, but I have a few questions:

1. How does this differ from other revenue-sharing systems, e.g. NFTs on RMRK which give a percentage of every sale to the initial creator?
2. Let's say that I build on top of another system. How is the percentage determined? What stops me from setting it to 0% (or 0.0001%) or something so I am not giving up anything (or not much of anything)?

hoho-yin commented:
> Thanks for the update and the additional information. I will mark the application as ready for review, so that others start looking into it as well. But could you explain for example why you need to create the token standard, since you removed the part about “shares”?

Dear @Noc2 :
Hello!

Thanks for your kindly marking, and welcome others to review and give any golden critiques. 

Although removing "shares" to not distribute revenue within each work, we think distributing revenue between new work and its referred old works are needed. So token is used to record the reference relationship among different works for the revenue distribution between new work and all its referred works.

Any other questions or feedback, feel free to put forward. Thanks a lot and hope can get your endorsement!

Kindly Regards,
Jieliang Yin

semuelle commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+
+3. A revenue allocation and distribution system that confirms the revenue distribution scheme agreed by different works with the reference relationship. It will track the earnings of the new work, and distribute them with the old work which the new work refers to according to their distribution scheme.
+
+All programs will be developed in !ink on Pallets.
```
Smart contracts don't require any pallets, other than the `contracts` pallet. I suppose you mean "for parachains with `contracts` pallet?

semuelle commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+| 0e. | Article | We will publish an article that explains the design and usage of our work in details. |
+|  1. | Specification | We will specify the details of our token standard design that how users can create their own work, refer to others' work, and get the earnings of their work via the reference token in the reference system. |
+|  2. | Setup | We provide a repository that forms the basis of our deliverables. |
+|  3. | Core Types | We provide the core data and coding types in details. |
```
Can you be more specific here? Which types? Is this a full implementation or just a specification?

hoho-yin replied:
Yes we will fully implement it in our demo. Core types mean the data types in the objects of our token standard. Will revise it more clearly.

semuelle commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | We will design a new reference license in our project, which records and reflects the reference relationship between the referred old works and the new works that are based on the old works. |
+| 0b. | Documentation | We will provide both inline documentation of the code and a basic tutorial that explains how a user can spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
```
I strongly suggest using substrate-contracts-node here.

```suggestion
| 0b. | Documentation | We will provide both inline documentation of the code and a basic tutorial that explains how a user can spin up a [Substrate contracts node](https://github.com/paritytech/substrate-contracts-node), deploy and intialize the contracts and send test transactions, which will show how the new functionality works. |
```

semuelle commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+| 0e. | Article | We will publish an article that explains the design and usage of our work in details. |
+|  1. | Specification | We will combine all our previous works above together to build the integrated system. We will specify the configurations and implementations of the whole system, and encourage everyone within the community to use, test and improve. |
+|  2. | Setup | We provide a repository that forms the basis of our deliverables. |
+|  3. | Core Types | We provide the core data and coding types in details. |
```
see above

semuelle commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+|  1. | Specification | We will combine all our previous works above together to build the integrated system. We will specify the configurations and implementations of the whole system, and encourage everyone within the community to use, test and improve. |
+|  2. | Setup | We provide a repository that forms the basis of our deliverables. |
+|  3. | Core Types | We provide the core data and coding types in details. |
+|  4. | User Testing | We will make the testing plan and invite users for testing and evaluation. After that we will produce testing report, and improve our system continually based on that. |
```
Can you be more specific here? What will you be measuring? Who will be your cohort?

hoho-yin replied:
We will mainly measure the degree and utility of decentralization in Web3 after implementing our project. How decentralized Web3 could be after using our system, and how online services could be better on such more decentralized Web3. We will also test the functionalities, security, and performance of the demo we built for our system. The cohort will be the Web3 users.

semuelle commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+
+Therefore Web3 aims at solving the problem of non-incentive in Web1, and centralization in Web2 to make online services both decentralized and incentive. In Web3, online public services should be maintained and managed not by any centralized entity in private, but by users in public collectively. Users will not only consume and provide online services, but also manage them together to protect the ...
+
+Some Web3 projects try to break the monopoly among different platforms, and give some rights of self-governance to the community via DAO, but still far from the goal above. A central team like the founding company still has more privileges and plays a more vital role than others in determining the development and management of its project, and they predefine and limit the self-governance of the D ...
```
That is a very narrow view of what DAOs today are. Whether tokens are required for participation and how power is distributed among its members is a design decision, not a technical limitation. There are DAO platforms in existence that let you design and deploy DAOs with whatever mechanics you desire.

hoho-yin replied:
Yes we are exploring a new mechanism of DAO besides staking and voting. DAO could be designed in any way without technical limitation, so we are designing a new way and we think it could be better than any other existing ways.

semuelle reviewed (changes requested):
I have trouble following this application. Please have a look at my questions below. Also, please move the .md file to the `applications` folder, and move the images elsewhere. If you drop an image into the Github editor, it will automatically upload the file to `githubusercontent.com`, that'd be ideal.

hoho-yin commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+
+- **Project Name:** Towards fully decentralized and incentive Web3 by designing a non-fungible reference blockchain system
+- **Team Name:** Jieliang Yin
+- **Payment Address:** Will be given after the application has been approved
```
All work for us like BTC or ETH

hoho-yin commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+| -----: | ----------- | ------------- |
+| 0a. | License | We will design a new reference license in our project, which records and reflects the reference relationship between the referred old works and the new works that are based on the old works. |
+| 0b. | Documentation | We will provide both inline documentation of the code and a basic tutorial that explains how a user can spin up one of our Substrate nodes and send test transactions, which will show how the new functionality works. |
+| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
```
DAO community is not a program, but something on Discord or Telegram like other community of Web3 projects. We will define the governance rules and operation plan of our community as deliverables in this milestone. We will also determine where it would be, whether building a website for that or using existing tools. Community is very important for our project, and can undertake several roles like filtering different versions of one work, and defending the copyright of one work.

hoho-yin commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+
+### Ecosystem Fit
+
+![img](https://github.com/hoho-yin/Grants-Program/blob/hoho-yin-patch-1/impact.png)
```
In the Use Case part we take Uniswap as example. Uniswap only allows one version at one time, like v1 in 2018, v2 in 2020,   and v3 in 2021, and they are all made by the official. In our reference system, instead we allow different versions at one time, like v1.1, v1.2, v1.3... in 2018, v2.1, v2.1.1, v2.1.1.1 in 2020... and they are made by different people not only the official team. Although made by different people, different versions can access and interoperate with the official version v1, v2, or v3, like soft fork not hard fork.

hoho-yin commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+
+For example, currently one DeFi application couldn’t have multiple versions made by different people. If others change a DeFi protocol to make a new version of it, it couldn’t get the internal network and resources of the original version. For instance, Sushiswap improved Uniswap, but Uniswap couldn’t open and share its liquidity pool with Sushiswap. Sushiswap has to build a new protocol complete ...
+
+With our reference system, Uniswap can have unlimited new versions just like branches or forks made by anyone including Sushiswap. Uniswap opens and shares its internal users and network and allows anyone to make any changes on itself, so Sushiswap can access the liquidity pools of Uniswap and build their protocol based on them directly with no need to build their own from scratch. In return, the ...
```
We think Web3 services should be fully decentralized. The development of each service should be out of owner's control. Owner should not control and determine the future of the service like roadmap following his intention, but should allow different people can improve and change it freely as long as markets and users need, even beyond owner's scope. 

For example although Uniswap doesn't intend to have governance token in the beginning, it should still allow others like Sushiswap to add platform token in the new version developed by Sushiswap. Users can freely choose old or new version. Old and new version can share the same liquidity pool, so Sushiswap can directly build their project based on Uniswap rather than having to build a new project from scratch and launching vampire attack to scramble the liquidity of Uniswap. Such collaboration rather than vicious competition would be beneficial to both Uniswap and Sushiswap.

hoho-yin commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+| Append | address,token_id,reference,content  | Can’t modify original items |  |
+| Transfer | address,token_id | Change the ownership of one’s token | address, transaction |
+
+When creating a new work, a corresponding token will be minted automatically with unique token id to represent the new work and store all its referring works. The token also defines when it's referred by the following new tokens, how the earnings of the new work should be distributed with it.
```
We want something to represent the reference relationship between old work and the new work based on the old work. No existing token supports such reference function, not to mention distributing the revenue of the new work with its referred old work. So the new token we design allow new work to refer to the old work, build new work based on the referred old work, and distribute revenue with the old work.

hoho-yin reviewed (commented):
Changes done

Noc2 commented on Grant Application - Reference blockchain system.md:
```diff
@@ -0,0 +1,286 @@
+| 0d. | Docker | We will provide a Dockerfile that can be used to test all the functionality delivered with this milestone. |
+| 0e. | Article | We will publish an article that explains the design and usage of our work in details. |
+|  1. | Specification | We will specify the details of our token standard design that how users can create their own work, refer to others' work, and get the earnings of their work via the reference token in the reference system. |
+|  2. | Setup | We provide a repository that forms the basis of our deliverables. |
```
@semuelle points out that “We provide a repository that forms the basis of our deliverables.” isn’t necessary.

Noc2 reviewed (changes requested):
Something to be aware of: your current application only lists “specification” as deliveries and no milestone actually mentions the implementation as ink! Smart contracts or pallets. Could you add this?

laboon commented:
> > 1. How does this differ from other revenue-sharing systems, e.g. NFTs on RMRK which give a percentage of every sale to the initial creator?
> 
> Yes some NFTs allow others to use their work to create new work like Yuga Labs, but such right is limited to only the owner of the NFT. In our system instead everyone can build their new work based on any work no matter who you are, and do whatever changes you want without any limitation as long as referring to the work you based on.
> 
> Another difference is that current NFT focuses on the ownership of a digital asset, but in our reference system each token just represents the work nominally, and the token holder just has the shares of the earnings of the work like stocks. You won't have the actual ownership of a company if buying its stocks, just owing it virtually and sharing its dividend. You can also develop and improve the referred work just like shareholders participating in the management of a company.

OK, thank you for the explanation.


> In general there is a game theory between new work and the referred old work to make agreement on the final proportion of revenue distribution fairly. If the old work asks too much from the new work, no new work would refer to it, and the old work would get nothing. If the old work earns too little from the new work, it would be unwilling to open and share its work, and the new work would get nothing. Finally they could reach an equilibrium, and we are doing research on how to make such negotiation more automatic, efficient and reliable, which would also be our project outcome.

Also interesting, glad you have thought of this.

I'll be checking the responses to the other requests made by other members of the team on this PR.

hoho-yin commented:
> pallets

Dear @Noc2 
Hello!

Thanks for your kindly review and feedback. We have added specific deliverables of implementation work into Milestone coded via ink! and implemented on Pallets. 

Any other questions, feel free to put all of them forward at one time, so you could give your final judgement for our proposal after our final revision. Thanks a lot for your kindly support!


Kindly Regards,
Jieliang Yin

hoho-yin commented:
> OK, thank you for the explanation.
> 
> 
> Also interesting, glad you have thought of this.
> 
> I'll be checking the responses to the other requests made by other members of the team on this PR.

Dear @laboon :
Hello!

Thanks for your kindly revision and affirmation. As you have no other questions, could you please make final judgment on our proposal like endorsing our grant application or not? Thanks a lot for your kindly support!

Kindly Regards,
Jieliang Yin

Noc2 reviewed (changes requested):
Thanks for adding ink! smart contracts to the grant application. I told the rest of the team to take another look at the application. However, I’m personally still not convinced. I think the general idea is great, but the application is missing the concrete technical details on how to archive this. For example, you mention security and privacy, but don’t explain how you are actually going to implement it. Or you mention that everyone can participate in the DAO, but how does this work? Someone needs to pay for the on-chain fees. Also don’t you need some kind of oracle for your system to work?

semuelle commented:
I will not support this project one way or another. I think the downsides of integrating with such a system are larger than the upsides. Also, the technical specification is at times imprecise, at times confusing. This project is like a standardization attempt; to be at least somewhat successful, you will probably have to start with a consortium of big projects on your side.

hoho-yin commented:
@semuelle @Noc2 
The details of the project will be specified after starting the development and implementation. I couldn't determine them now. It's impractical to ask for concrete details at this point. As for your other opinions, they have no ground to support them. You didn't give any reason for your judgement, so I don't know how to revise accordingly. Due to your unreasonable requirements and illogical judgments, I have to apply for other grants. Thanks a lot!
